### ISSUE TITLE
Items that have been removed from cache trigger onExpire

### ISSUE THREAD

**Issue by vlad-zhukov** [planttheidea/moize#89]
I am not sure what would be the best way to fix it, so I've just created a test. Please also note a case when one removes a key from the cache and adds the same key back like in a test below.

```js
test('if moize will not call onExpire for removed cache items', async (t) => {
  const fn = sinon.spy();
  const options = {
    maxAge: 500,
    onExpire: sinon.spy()
  };

  const moized = moize(fn, options);

  const args = ['foo', 'bar', 'baz', 'quz'];

  moized(...args);

  t.is(moized.keys().length, 1);

  await new Promise((resolve) => {
    setTimeout(resolve, 200);
  });

  moized.remove(args);
  moized(args);

  t.is(moized.keys().length, 1);

  await new Promise((resolve) => {
    setTimeout(resolve, 400);
  });

  t.is(options.onExpire.callCount, 1); // ok

  await new Promise((resolve) => {
    setTimeout(resolve, 200);
  });

  t.is(options.onExpire.callCount, 1); // error, got 2
});
```

**Comment by planttheidea** [comment]
@vhfmag thanks for this, and the test was very helpful. The solution was very straightforward, I just needed to clear the timeout of the expiration and remove it from the list like I already do upon expiration.

I'll have a PR shortly.

**Comment by planttheidea** [comment]
So maybe I'm misunderstanding, because based on the description the test does not align with the problem.

My interpretation of the problem was that removed keys were still firing `onExpire`, and my change prevents that from happening. Your test appears to still expect `onExpire` to fire, albeit in a weird way. 

First, when you "reapply" the key, you arent passing the same key ... you are using `moized(args)` when it should be `moized(...args)`, as in the initial call. Second, you are expecting `onExpire` to fire after `400` milliseconds, when your `maxAge` is `500`. When you removed the keys, the original `onExpire` was cleared and removed, and when you "reapply" the same keys, a new expiration is created. Try these tweaks:

```javascript
test('if moize will not call onExpire for removed cache items', async (t) => {
  const fn = sinon.spy();
  const options = {
    maxAge: 500,
    onExpire: sinon.spy()
  };

  const moized = moize(fn, options);

  const args = ['foo', 'bar', 'baz', 'quz'];

  moized(...args);

  t.is(moized.keys().length, 1);

  await new Promise((resolve) => {
    setTimeout(resolve, 200);
  });

  moized.remove(args);

  t.is(moized.keys().length, 0); // keys are gone

  await new Promise((resolve) => {
    setTimeout(resolve, options.maxAge);
  });

  t.true(options.onExpire.notCalled); // good, it never fired

  moized(...args); // reapply the same keys

  t.is(moized.keys().length, 1); // good, the key is back

  await new Promise((resolve) => {
    setTimeout(resolve, options.maxAge + 50); // a little longer than the maxAge just to be safe
  });

  t.true(options.onExpire.calledOnce); // good, it fired only the one time we expected it to
});
```

If this does not align with your expectations, let me know.

**Comment by planttheidea** [comment]
Haha so this was just me being stupid, mainly because I didn't bother to include the integration test you were kind enough to provide in the final product! Note to self: unit tests are not enough.

Basically, when trying to remove the expiration I was using the exact structure of the key to remove ... but not the key itself. The utility to find the expiration index therefore could not find the key, so it didn't actually do anything. I changed the `remove` function to instead use the real key from the existing cache (captured before the key is removed), and voila ... it works as intended.

I'll get this out in the next few minutes, with the above integration test included to prove it! :)

**Comment by planttheidea** [comment]
Should actually be fixed in `5.2.3.`! If you are still experiencing issues with this let me know, but I think we are in the clear now. Appreciate it!

**Comment by vlad-zhukov** [comment]
Thanks! 🎉
